feat: issue_03

yoshikawa-akr opened the pull request:
# [issue No.3](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#3)

**issue**

このDOM
```
<ul>
  <li><a href="a1.html"><img src="/img/bookmark.png">a1</a></li>
  <li><a href="a2.html"><img src="/img/message.png">a2</a></li>
</ul>
```
をJavaScriptでつくり、html内のulの中に差し込んでください

## [CodeSandBox](https://codesandbox.io/s/laughing-sunset-ctxi40) / Latest commit: https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/21fe2815c83b383b40c3390ec092bcf16b13592c

## In terms of implementation, what you are worried about, what you want to focus on, and what you want to ask.
- 92d0e4dc5660a852e44a2de96c3f95b46d7920a1
I changed the attributes and text from the duplicated element, but I'm not sure if I can create a similar element more efficiently.

yoshikawa-akr committed:
feat: Get the ul element.

Give the ul element an id and get it with getElementById.

yoshikawa-akr committed:
feat: Create new li,a,img tags.

yoshikawa-akr committed:
Add new attributes and text.

yoshikawa-akr committed:
feat: Specify a child element of ul.

Insert li, a, and img tags as child elements of ul at the specified positions.

yoshikawa-akr committed:
feat: Duplicate elements, change attributes.

Duplicate the li tag and its child elements, and change the attributes. Then insert the duplicated li element anywhere you want.

yoshikawa-akr committed:
style: Delete blank lines.

csakyo commented on issue_03/script.js:
```diff
@@ -0,0 +1,18 @@
+
+nodeListOfCloneA[0].src = '/img/message.png';
+nodeListOfCloneA[1].textContent = 'a2';
+li.after(cloneLi);
```
I did not know the "cloneNode" !!:eyes: 

Q and IMO
I felt that if I wanted to repeat the same thing 10 times, I would write the same description 10 times if I used this method. I thought it would be better to use a more efficient method.
(I do not know this method well so I would like to know your opinion! )

yoshikawa-akr replied:
Thank you for your comment.
It is true that if you need 10 clones, you will need to make the same statement 10 times.
For this reason, I modified the code to create an array of attributes and text, and handle the creation and insertion of elements in a loop.
[d9a0d6a](https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/d9a0d6a9f687c8404873337811d0c3b847aacea8)

yoshikawa-akr committed:
refactor: Changed to loop processing.

Create a new array of attribute values and text, create li, a, and img tags, specify attributes and text, and loop through insertion at any point.

yoshikawa-akr commented:
> How about using iteration for similar items? For example, the "a" and "img" parts. Why not use an array for "bookmark.png" and "message.png"?

Thank you for your comment.
You are right, looping is a better way to create something similar.
I created an array for attributes and text and used a loop to handle the creation and insertion of elements.
[d9a0d6a](https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/d9a0d6a9f687c8404873337811d0c3b847aacea8)

csakyo commented on issue_03/script.js:
```diff
@@ -0,0 +1,16 @@
+    a.textContent = textArray[i];
+    img.setAttribute('src', srcArray[i]);
+    a.insertAdjacentElement('afterbegin', img);
+    ul.appendChild(li).appendChild(a);
```
IMO
You can even write it like this!
```suggestion
ul.appendChild(li).appendChild(a).insertAdjacentElement('afterbegin', img);
```

yoshikawa-akr replied:
I must have missed that!
Thanks for pointing it out :+1:
[aae4ba6](https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/aae4ba6b09214ad5dd7f5facd35a829aaae31f71)

csakyo commented on issue_03/script.js:
```diff
@@ -0,0 +1,16 @@
+const textArray = ['a1', 'a2'];
+const srcArray = ['/img/bookmark.png', '/img/message.png'];
+
+for (let i = 0; i < 2; i++) {
```
IMO
This ``i < 2`` is a magic number that I don't know what 2 is, so an explanatory variable might be a good idea.

For example, 
``const liLength = 2;`` 
``for (let i = 0; i < liLength ; i++){...``

Then it's easy to imagine what number 2 is.

OR you can also use arrays I think.
``for (let i= 0; i < srcArray.length; i++){...``

yoshikawa-akr replied:
I combined the arrays into one.
[7e04e9a](https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/7e04e9a66be55ed691a301bfe81d4db8e2959e70)

The condition for looping is set using the length property of the array.
[d4100b5](https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/d4100b59da7a372febb91d1d54b73551610f3fb0)

csakyo commented on issue_03/script.js:
```diff
@@ -0,0 +1,16 @@
+    a.textContent = textArray[i];
+    img.setAttribute('src', srcArray[i]);
+    a.insertAdjacentElement('afterbegin', img);
+    ul.appendChild(li).appendChild(a);
+}
```
IMO
Using appendchild in a for loop is not good for performance.
You can study "createDocumentFragment" here! This is good study so please refer the Monaka san's article!

https://monatamalog.com/archives/291

yoshikawa-akr replied:
I see that the performance is not degraded by having only one reflow for batch insertion.
I learned a lot!!!
Thank you very much :raised_hands:
https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/eb7360804f572e0a1c3c5fe37fc11ac6a3ea19e8

seeeen044 commented on issue_03/script.js:
```diff
@@ -0,0 +1,16 @@
+const textArray = ['a1', 'a2'];
+const srcArray = ['/img/bookmark.png', '/img/message.png'];
+
+for (let i = 0; i < 2; i++) {
```
IMO
```
const array = [
  { href: "a1.html", src:"/img/bookmark.png",text: "a1"},
  { href: "a2.html", src:"/img/message.png",text: "a2"}
]
```
Put together the array.

`for (let i = 0; i < array.length ; i++){`
to avoid magic numbers and improve the readability of the code.

yoshikawa-akr replied:
The array is combined into one.
It's a concept I never had, so I learned a lot!
Thank you very much :raised_hands:
https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/7e04e9a66be55ed691a301bfe81d4db8e2959e70
https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/d4100b59da7a372febb91d1d54b73551610f3fb0

seeeen044 commented on issue_03/script.js:
```diff
@@ -0,0 +1,16 @@
+    const a = document.createElement('a');
+    const img = document.createElement('img');
+
+    a.setAttribute('href', hrefArray[i]);
+    a.textContent = textArray[i];
+    img.setAttribute('src', srcArray[i]);
```
```suggestion
    a.href = array[i].href;
    a.textContent = array[i].text;
    img.src = array[i].src;
```
nits
You can also write it this way.

yoshikawa-akr replied:
Thank you for pointing that out.
It's fixed :sparkles:
https://github.com/yoshikawa-akr/js-lessons/pull/3/commits/21fe2815c83b383b40c3390ec092bcf16b13592c

yoshikawa-akr committed:
refactor: Creating an associative array.

Three arrays combined into one associative array.

yoshikawa-akr committed:
refactor: Changing conditional expressions in loops and changing references during processing.

Set contentsArray.length as the conditional expression.
Refer to contentsArray when specifying attributes or text.

yoshikawa-akr committed:
refactor: Change to method chain.

yoshikawa-akr committed:
refactor: Creating a documentFragment.

yoshikawa-akr committed:
refactor: Changing the reference complement of an attribute.
